### ISSUE TITLE
how to avoid empty foreground (null values of the softmax probabilities almost everywhere) when using SurfaceLoss alone

### ISSUE THREAD

**Issue by kkhuang1990** [LIVIAETS/boundary-loss#20]
hi, liviaets, 
thank you so much for sharing your codes.
I throughly read your paper and tried to run your codes with only surface loss while i met the same problem of obtaining empty foreground. Of course relatively lower gradient is one cause. I am wondering whether you have some ideas on avoiding this instead of combining the surface loss with another loss. Besides, i noticed that you consider probabilities of all pixels in predicted probability map which is a little different from the mathematical analysis, where only G\S and S\G are considered. probably this problem can be solved if we only consider probabilities corresponding to G\S and S\G

**Comment by kkhuang1990** [comment]
i also applied the surface  loss to multiple class segmentation problem while the performance is not as good as i desire. Especially after the surface loss ratio becomes 1.0, the performance begins to drop dramatically. I am guessing that's because the gradient learning is not stable when using surface loss alone.

**Comment by HKervadec** [comment]
Hey,

> I am wondering whether you have some ideas on avoiding this instead of combining the surface loss with another loss.

Somewhat pre-trained network or multi-class (more than 2 classes) problems ; I successfully made it work by itself on [ACDC](https://acdc.creatis.insa-lyon.fr/).

> Besides, i noticed that you consider probabilities of all pixels in predicted probability map which is a little different from the mathematical analysis, where only G\S and S\G are considered. probably this problem can be solved if we only consider probabilities corresponding to G\S and S\G

Yes but the initial problem is that you cannot define S (in a differentiable way) ; hence why we reword it with indicator functions, that we then relax with the softmax. Issue https://github.com/LIVIAETS/surface-loss/issues/9#issuecomment-488840311 might help you on that.

> i also applied the surface loss to multiple class segmentation problem while the performance is not as good as i desire. Especially after the surface loss ratio becomes 1.0, the performance begins to drop dramatically. I am guessing that's because the gradient learning is not stable when using surface loss alone. 

Ah that is interesting ; I didn't make many experiments with those settings. On ACDC the perfs were similar or slightly lower that a default cross-entropy, but I don't think I saw a drop in performances. Analyzing that would be interesting. Can you tell me what problem it was ?

**Comment by kkhuang1990** [comment]
```
Yes but the initial problem is that you cannot define S (in a differentiable way) ; hence why we reword it with indicator functions, that we then relax with the softmax. Issue #9 (comment) might help you on that.
``` 
we can obtain S by argmax and then detach it from the probability map output. In this way, we can define S\G and G\S, and further calculate surface loss as the mathematical equation shows. In this way, we can avoid the in-differentiable problem.

**Comment by HKervadec** [comment]
> we can obtain S by argmax and then detach it from the probability map output.

Yes, and then you could compute it as a _metric_, but you wouldn't be able to back-propagate on that, since it is non-derivable. 

A loss requires to be derivable, while a metric doesn't. It is true that the formulation that we propose (or its discrete equivalent) could be used as a metric ; but I am not sure of its usefulness.
